src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,312 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ObjectMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\ObjectMapper\Exception\MappingException;
+use Symfony\Component\ObjectMapper\Exception\MappingTransformException;
+use Symfony\Component\ObjectMapper\Metadata\Mapping;
+use Symfony\Component\ObjectMapper\Metadata\ObjectMapperMetadataFactoryInterface;
+use Symfony\Component\ObjectMapper\Metadata\ReflectionObjectMapperMetadataFactory;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @experimental
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class ObjectMapper implements ObjectMapperInterface
+{
+    /**
+     * An SplObjectStorage that tracks recursive references.
+     */
+    private ?\SplObjectStorage $objectMap = null;
+
+    /**
+     * @param ContainerInterface $transformCallableLocator
+     * @param ContainerInterface $conditionCallableLocator
+     */
+    public function __construct(
+        private readonly ObjectMapperMetadataFactoryInterface $metadataFactory = new ReflectionObjectMapperMetadataFactory(),
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $transformCallableLocator = null,
+        private readonly ?ContainerInterface $conditionCallableLocator = null,
+    ) {
+    }
+
+    public function map(object $source, object|string|null $target = null): object
+    {
+        $objectMapInitialized = false;
+        if (null === $this->objectMap) {
+            $this->objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $metadata = $this->metadataFactory->create($source);
+        $map = $this->getMapTarget($metadata, null, $source);
+        $target ??= $map?->target;
+        $mappingToObject = \is_object($target);
+
+        if (!$target) {
+            throw new MappingException('Mapping target not found.');
```

Comment:
Maybe we should add the source class in all error messages where it makes sense?